tests/broadcast.rs

```diff
@@ -1,164 +1,236 @@
     let _ = env_logger::try_init();
 
     // Make node 0 propose the value.
-    network.input(NodeId(0), proposed_value.to_vec());
+    let _step = net
+        .send_input(0, proposed_value.to_vec(), rng)
+        .expect("Setting input failed");
 
-    // Handle messages in random order until all nodes have output the proposed value.
-    while !network.nodes.values().all(TestNode::terminated) {
-        network.step();
+    // Handle messages until all good nodes have terminated.
+    while !net.nodes().all(|node| node.algorithm().terminated()) {
+        let _ = net.crank_expect(rng);
     }
-    // Verify that all instances output the proposed value.
-    for node in network.nodes.values() {
-        assert!(once(&proposed_value.to_vec()).eq(node.outputs()));
-    }
-    assert!(once(&proposed_value.to_vec()).eq(network.observer.outputs()));
-}
 
-fn new_broadcast(netinfo: Arc<NetworkInfo<NodeId>>) -> Broadcast<NodeId> {
-    Broadcast::new(netinfo, NodeId(0)).expect("Instantiate broadcast")
+    // Verify that all instances output the proposed value.
+    assert!(net
+        .nodes()
+        .all(|node| once(&proposed_value.to_vec()).eq(node.outputs())));
 }
 
-fn test_broadcast_different_sizes<A, F>(new_adversary: F, proposed_value: &[u8])
-where
+fn test_broadcast_different_sizes<A, F>(
+    new_adversary: F,
+    proposed_value: &[u8],
+    adversary_netinfo: &Arc<Mutex<NetworkInfoMap>>,
+) where
     A: Adversary<Broadcast<NodeId>>,
-    F: Fn(BTreeMap<NodeId, Arc<NetworkInfo<NodeId>>>) -> A,
+    F: Fn() -> A,
 {
     let mut rng = rand::thread_rng();
     let sizes = (1..6)
         .chain(once(rng.gen_range(6, 20)))
         .chain(once(rng.gen_range(30, 50)));
     for size in sizes {
+        // cloning since it gets moved into a closure
+        let cloned_netinfo_map = adversary_netinfo.clone();
         let num_faulty_nodes = util::max_faulty(size);
-        let num_good_nodes = size - num_faulty_nodes;
         info!(
             "Network size: {} good nodes, {} faulty nodes",
-            num_good_nodes, num_faulty_nodes
+            size - num_faulty_nodes,
+            num_faulty_nodes
         );
-        let adversary = |adv_nodes| new_adversary(adv_nodes);
-        let network = TestNetwork::new(num_good_nodes, num_faulty_nodes, adversary, new_broadcast);
-        test_broadcast(network, proposed_value);
+
+        let (net, _) = NetBuilder::new(0..size as u16)
+            .num_faulty(num_faulty_nodes as usize)
+            .message_limit(10_000 * size as usize)
+            .no_time_limit()
+            .adversary(new_adversary())
+            .using(move |info| {
+                let netinfo = Arc::new(info.netinfo);
+                cloned_netinfo_map
+                    .lock()
+                    .unwrap()
+                    .insert(info.id, netinfo.clone());
+                Broadcast::new(netinfo, 0).expect("Failed to create a ThresholdSign instance.")
```

Comment:
Should that say `Broadcast`? (Also below.)

Reply:
Fixed